### PR TITLE
[PTD][DCP] Add fsdp checkpoint example

### PR DESCRIPTION
**Pull request by wz337**
Add an example to show recommended way to checkpoint FSDP.

### CODE REVIEW

File: torch/distributed/checkpoint/examples/fsdp_checkpoint_example.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    torch.cuda.set_device(rank)
+
+    # Create a model
+    model_1, optim_1 = init_model()
+
+    # Save the model to CHECKPOINT_DIR
+    with FSDP.state_dict_type(model_1, StateDictType.SHARDED_STATE_DICT):
+        state_dict = {
+            "model": model_1.state_dict(),
+            "optim": FSDP.sharded_optim_state_dict(model_1, optim_1),
```
Comment (fegin) [inline torch/distributed/checkpoint/examples/fsdp_checkpoint_example.py:75]: Change this to `FSDP.optim_state_dict`.

File: torch/distributed/checkpoint/examples/fsdp_checkpoint_example.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        )
+        model_2.load_state_dict(state_dict["model"])
+
+        optim_state = load_sharded_optimizer_state_dict(
+            model_state_dict=state_dict["model"],
+            optimizer_key="optim",
+            storage_reader=dist_cp.FileSystemReader(CHECKPOINT_DIR),
+        )
+
+        flattened_osd = FSDP.flatten_sharded_optim_state_dict(
```
Comment (fegin) [inline torch/distributed/checkpoint/examples/fsdp_checkpoint_example.py:109]: Change this to `FSDP.optim_state_dict_to_load`.

**Review by fegin — CHANGES_REQUESTED** [review]

**Review by kumpera — APPROVED** [review]
This looks great.

File: torch/distributed/checkpoint/examples/fsdp_checkpoint_example.py (hunk 2 above)
Comment (kumpera) [reply torch/distributed/checkpoint/examples/fsdp_checkpoint_example.py]: do we need to use the context manager here?

**Pull request closed without merging**